### PR TITLE
Add ignore_hyperparameters to MLFlowLogger

### PR DESCRIPTION
**Pull request by ngcgarcia**
# What does this PR do?

Allows users to skip logging certain hyperparameters specified by input patterns.

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: tests/loggers/test_mlflow_logger.py [hunk 1]
```diff
@@ -642,3 +642,59 @@ def test_mlflow_ignore_metrics(tmp_path, device):
...
+    eval_interval = '1ba'
+
+    trainer = Trainer(model=SimpleConvModel(),
+                      loggers=test_mlflow_logger,
+                      train_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
+                      eval_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
+                      max_duration=f'{num_batches}ba',
+                      eval_interval=eval_interval,
+                      device=device)
+    trainer.fit()
```
Comment (dakinggg) [inline tests/loggers/test_mlflow_logger.py:671]: do you need to call fit? If not, please remove.

File: tests/loggers/test_mlflow_logger.py [hunk 2]
```diff
@@ -642,3 +642,59 @@ def test_mlflow_ignore_metrics(tmp_path, device):
...
+def test_mlflow_ignore_hyperparameters(tmp_path, device):
+    mlflow = pytest.importorskip('mlflow')
+
+    mlflow_uri = tmp_path / Path('my-test-mlflow-uri')
+    experiment_name = 'mlflow_logging_test'
+    test_mlflow_logger = MLFlowLogger(tracking_uri=mlflow_uri,
+                                      experiment_name=experiment_name,
+                                      ignore_hyperparameters=['num*', 'mlflow_run_id', 'nothing'])
+    # Reduce the system metrics sampling interval to speed up the test.
+    mlflow.set_system_metrics_sampling_interval(1)
```
Comment (dakinggg) [inline tests/loggers/test_mlflow_logger.py:657]: lets just not do system metrics in this test? and get rid of the sleep call too?

File: tests/loggers/test_mlflow_logger.py [hunk 3]
```diff
@@ -640,5 +636,49 @@ def test_mlflow_ignore_metrics(tmp_path, device):
...
+                                      ignore_hyperparameters=['num*', 'mlflow_run_id', 'nothing'])
+
+    dataset_size = 64
+    batch_size = 4
+    num_batches = 4
+    eval_interval = '1ba'
+
+    Trainer(model=SimpleConvModel(),
+            loggers=test_mlflow_logger,
+            train_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
```
Comment (dakinggg) [inline tests/loggers/test_mlflow_logger.py:656]: you can just make the dataloaders empty lists too, since you're not using them for anything.

**Review by dakinggg — APPROVED** [review]

File: composer/loggers/mlflow_logger.py [hunk 4]
```diff
@@ -56,6 +56,7 @@ class MLFlowLogger(LoggerDestination):
         log_system_metrics (bool, optional): Whether to log system metrics. If ``True``, Mlflow will
             log system metrics (CPU/GPU/memory/network usage) during training. (default: ``True``)
         ignore_metrics (List[str], optional): A list of glob patterns for metrics to ignore when logging. (default: ``None``)
+        ignore_hyperparameters (List[str], optional): A list of glob patterns for hyperparameters to ignore when logging. (default: ``None``)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:59]: Thought: what if we just use `ignore_metrics` in both places? I'm good either way though

File: tests/loggers/test_mlflow_logger.py [hunk 5]
```diff
@@ -640,5 +636,49 @@ def test_mlflow_ignore_metrics(tmp_path, device):
     metric_file = run_file_path / Path('metrics') / Path('system/cpu_utilization_percentage')
     assert not os.path.exists(metric_file)
 
-    # Undo the setup to avoid affecting other test cases.
-    mlflow.set_system_metrics_sampling_interval(None)
+
+@device('cpu')
+def test_mlflow_ignore_hyperparameters(tmp_path, device):
```
Comment (mvpatel2000) [inline tests/loggers/test_mlflow_logger.py:641]:
```suggestion
def test_mlflow_ignore_hyperparameters(tmp_path):
```
default is CPU

File: tests/loggers/test_mlflow_logger.py [hunk 6]
```diff
@@ -640,5 +636,49 @@ def test_mlflow_ignore_metrics(tmp_path, device):
...
+    eval_interval = '1ba'
+
+    Trainer(model=SimpleConvModel(),
+            loggers=test_mlflow_logger,
+            train_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
+            eval_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
+            max_duration=f'{num_batches}ba',
+            eval_interval=eval_interval,
+            device=device)
+    test_mlflow_logger.post_close()
```
Comment (mvpatel2000) [inline tests/loggers/test_mlflow_logger.py:661]:
```suggestion

    Trainer(model=SimpleConvModel(),
            loggers=test_mlflow_logger,
            max_duration=f'4ba')
    test_mlflow_logger.post_close()
```

File: tests/loggers/test_mlflow_logger.py [hunk 7]
```diff
@@ -640,5 +636,49 @@ def test_mlflow_ignore_metrics(tmp_path, device):
...
+
+    run = _get_latest_mlflow_run(
+        experiment_name=experiment_name,
+        tracking_uri=mlflow_uri,
+    )
+    run_id = run.info.run_id
+    experiment_id = run.info.experiment_id
+
+    run_file_path = mlflow_uri / Path(experiment_id) / Path(run_id)
```
Comment (mvpatel2000) [inline tests/loggers/test_mlflow_logger.py:670]:
```suggestion
    run_file_path = mlflow_uri / Path(run.info.experiment_id) / Path(run.info.run_id)
```

**Review by mvpatel2000 — APPROVED** [review]

File: composer/loggers/mlflow_logger.py (hunk 4 above)
Comment (dakinggg) [reply composer/loggers/mlflow_logger.py:59]: they have different requirements :) e.g. maybe i want to ignore all hparams (`*`) without ignoring all metrics

Comment (ngcgarcia) [reply composer/loggers/mlflow_logger.py:59]: it's a lil silly that we have to support both of these, but they do function as separate systems and could potentially have overlap in names, so leaving them separate is prob smart

**Pull request merged**